File: Sorting/Exchange_Sorting/Cocktail_Sort/python/sort.py
from typing import List


class CocktailSort:

    def __init__(self, target_list: List[int]):
        self._target_list = target_list
        self._list_len = len(self._target_list)


    def sort(self) -> List:
        left_index = 0
        right_index = self._list_len - 1
        final_index = 0

        while left_index < right_index:
            # Sort the elements of list from left to right
            for _i in range(left_index, right_index):
                if self._target_list[_i] > self._target_list[_i+1]:
                    self.swap(_i, _i+1)
                    final_index = _i

            right_index = final_index

            # Sort the elements of list from right to left
            for _j in range(right_index, left_index, -1):
                if self._target_list[_j - 1] > self._target_list[_j]:
                    self.swap(_j, _j - 1)
                    final_index = _j

            left_index = final_index

        return self._target_list


    def swap(self, i: int, j: int) -> None:
        self._target_list[i], self._target_list[j] = self._target_list[j], self._target_list[i]


if __name__ == '__main__':

    _list = [7, 8, 9, 4, 7, 3, 5, 9, 8]

    print(f"Original list: {_list}")
    _insert_sort = CocktailSort(target_list=_list)
    print("Start to sort ...")
    _sorted_list = _insert_sort.sort()
    print(f"Sorted list: {_sorted_list}")

File: Sorting/Exchange_Sorting/Cocktail_Sort/c++/sort.cpp
//
//  sort.cpp
//  KobeXCodeCProject
//
//  Created by Bryant Liu on 2022/4/24.
//  Copyright © 2022 Bryant Liu. All rights reserved.
//

#include "sort.hpp"

#include <vector>
#include <iostream>

using namespace std;

CocktailSort::CocktailSort(vector<int> _list, int _len) {
    for (int i=0; i<_len; i++) list.push_back(_list[i]);
    this->len = _len;
}


void CocktailSort::cocktail_sort() {
    int i, left = 0, right = len - 1, shift = 0;
    
    while(left < right) {
        for (i = left; i < right; i++) {
            if (list[i] > list[i+1]) {
                swap(i, i+1);
                shift = i;
            }
        }
        
        right = shift;
        
        for (i = right; i > left; i--) {
            if (list[i] < list[i-1]) {
                swap(i, i-1);
                shift = i;
            }
        }
        
        left = shift;
    }
}


void CocktailSort::swap(int i, int j) {
    int temp = list[i];
    list[i] = list[j];
    list[j] = temp;
}


void CocktailSort::out() {
    for (auto const &i : list) {
        std::cout << i << std::endl;
    }
}


//int main(int argc, const char * argv[]) {
//    vector<int> myList= {4, 6, 2, 7, 5, 2, 4, 3, 6, 2, 3, 5, 6};
//
//    CocktailSort sort(myList, 13);
//    cout << "Before sorting" << endl;
//    sort.out();
//
//    cout << "Start to sort ..." << endl;
//    sort.cocktail_sort();
//
//    cout << "After sorting" << endl;
//    sort.out();
//
//    return 0;
//}

